roc.go

```diff
@@ -0,0 +1,110 @@
+
+import "sort"
+
+// ROC returns paired false positive rate (FPR) and true positive rate
+// (TPR) values corresponding to n cutoffs spanning the relative
+// (or receiver) operator characteristic (ROC) curve obtained when y is
+// treated as a binary classifier for classes with weights.
+//
+// Cutoffs are equally spaced from eps less than the minimum value of y
+// to the maximum value of y, including both endpoints -- this means the
+// resulting ROC curve will always begin at (0,0) and end at (1,1).
+//
+// The input y must be sorted, and SortWeightedLabeled can be used in
+// order to sort y together with classes and weights.
+//
+// For a given cutoff value, observations corresponding to entries in y
+// greater than the cutoff value are classified as false, while those
+// below (or equal to) the cutoff value are classified as true. These
+// assigned class labels are compared with the true values in the classes
+// slice and used to calculate the FPR and TPR.
+//
+// If weights is nil, all weights are treated as 1.
+//
+// n of zero results in all possible cutoffs being calculated -- this will
+// result in fpr and tpr having length len(y)+1 if y has no duplicated
+// values and one greater than the number of unique values in y in
+// general. n >= 2 will result in fpr and tpr having length n.
+// n of one or negative values of n are not supported.
+//
+// More details about ROC curves are available at
+// https://en.wikipedia.org/wiki/Receiver_operating_characteristic
+func ROC(n int, y []float64, classes []bool, weights []float64) (tpr, fpr []float64) {
+	if len(y) != len(classes) {
+		panic("stat: slice length mismatch")
+	}
+	if weights != nil && len(y) != len(weights) {
+		panic("stat: slice length mismatch")
+	}
+	if !sort.Float64sAreSorted(y) {
+		panic("stat: input must be sorted")
+	}
+
+	var incWidth, tol float64
+	if n != 0 {
+		if n < 2 {
+			panic("stat: cannot calculate fewer than 2 points on a ROC curve")
+		}
+		incWidth = (y[len(y)-1] - y[0]) / float64(n-1)
+		if incWidth <= 0 {
+			panic("stat: input constant")
+		}
+		tol = y[0] + incWidth
+		tpr = make([]float64, n)
+		fpr = make([]float64, n)
+	} else {
+		tpr = make([]float64, len(y)+1)
+		fpr = make([]float64, len(y)+1)
+	}
+	var bin int = 1 // the initial bin is known to have 0 fpr and 0 tpr
+	var nPos, nNeg, posWeight, negWeight float64
+	for i, u := range classes {
+
+		posWeight, negWeight = 0, 1
+		if weights != nil {
+			negWeight = weights[i]
+		}
+		if u {
+			posWeight, negWeight = negWeight, posWeight
+		}
+		nPos += posWeight
+		nNeg += negWeight
+		tpr[bin] += posWeight
+		fpr[bin] += negWeight
+
+		if n == 0 {
+			// Deal with duplicate y values
+			if i != (len(y)-1) && y[i] != y[i+1] {
+				bin++
+				tpr[bin] = tpr[bin-1]
+				fpr[bin] = fpr[bin-1]
+			}
+		} else {
+			// Deal with empty bins
+			for i != (len(y)-1) && y[i+1] > tol {
+				tol += incWidth
+				bin++
+				tpr[bin] = tpr[bin-1]
+				fpr[bin] = fpr[bin-1]
+			}
+		}
+	}
+	if n == 0 {
```

Comment:
Nope, my intention was that if n is specified and is not zero, then ROC should return slices of length n. So if there is a gap between one y value and the next that is greater than incWidth then that will result in consequtive entries in fpr and tpr that are equal -- so the same point on the ROC curve being repeated in the output. This makes sense to me because the concept of specifying n is that you are essentially specifying a list of n cutoffs to use, it just so happens that two (or more) of those cutoffs result in the same outcome. Does that make sense? The updated description of ROC should reflect that now.